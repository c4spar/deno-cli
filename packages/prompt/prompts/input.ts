import { blue } from 'https://deno.land/std@0.63.0/fmt/colors.ts';
import { Figures } from '../lib/figures.ts';
import { GenericInput, GenericInputKeys, GenericInputPromptOptions, GenericInputPromptSettings } from '../lib/generic-input.ts';

export interface InputKeys extends GenericInputKeys {}

export interface InputOptions extends GenericInputPromptOptions<string> {
    minLength?: number;
    maxLength?: number;
    keys?: InputKeys;
}

interface InputSettings extends GenericInputPromptSettings<string> {
    minLength: number;
    maxLength: number;
    keys?: InputKeys;
}

export class Input extends GenericInput<string, InputSettings> {

    public static async prompt( options: string | InputOptions ): Promise<string> {

        if ( typeof options === 'string' ) {
            options = { message: options };
        }

        return new this( {
            pointer: blue( Figures.POINTER_SMALL ),
            minLength: 0,
            maxLength: Infinity,
            ...options
        } ).prompt();
    }

    protected validate( value: string ): boolean | string {
        if ( typeof value !== 'string' ) {
            return false;
        }
        if ( value.length < this.settings.minLength ) {
            return `Value must be longer then ${ this.settings.minLength } but has a length of ${ value.length }.`;
        }
        if ( value.length > this.settings.maxLength ) {
            return `Value can't be longer then ${ this.settings.maxLength } but has a length of ${ value.length }.`;
        }
        return true;
    }

    protected transform( value: string ): string | undefined {
        return value.trim();
    }

    protected format( value: string ): string {
        return value;
    }
}
