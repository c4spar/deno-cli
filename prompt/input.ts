import { blue } from "./deps.ts";
import { Figures } from "./figures.ts";
import {
  GenericInput,
  GenericInputKeys,
  GenericInputPromptOptions,
  GenericInputPromptSettings,
} from "./_generic_input.ts";

export type InputKeys = GenericInputKeys;

/** Input prompt options. */
export interface InputOptions extends GenericInputPromptOptions<string> {
  minLength?: number;
  maxLength?: number;
  keys?: InputKeys;
}

/** Input prompt settings. */
interface InputSettings extends GenericInputPromptSettings<string> {
  minLength: number;
  maxLength: number;
  keys?: InputKeys;
}

/** Input prompt representation. */
export class Input extends GenericInput<string, InputSettings> {
  /** Execute the prompt and show cursor on end. */
  public static async prompt(options: string | InputOptions): Promise<string> {
    if (typeof options === "string") {
      options = { message: options };
    }

    return new this({
      pointer: blue(Figures.POINTER_SMALL),
      minLength: 0,
      maxLength: Infinity,
      ...options,
    }).prompt();
  }

  /**
   * Validate input value.
   * @param value User input value.
   * @return True on success, false or error message on error.
   */
  protected validate(value: string): boolean | string {
    if (typeof value !== "string") {
      return false;
    }
    if (value.length < this.settings.minLength) {
      return `Value must be longer then ${this.settings.minLength} but has a length of ${value.length}.`;
    }
    if (value.length > this.settings.maxLength) {
      return `Value can't be longer then ${this.settings.maxLength} but has a length of ${value.length}.`;
    }
    return true;
  }

  /**
   * Map input value to output value.
   * @param value Input value.
   * @return Output value.
   */
  protected transform(value: string): string | undefined {
    return value.trim();
  }

  /**
   * Format output value.
   * @param value Output value.
   */
  protected format(value: string): string {
    return value;
  }
}
