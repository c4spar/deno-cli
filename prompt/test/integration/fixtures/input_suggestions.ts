import { Input } from "../../../input.ts";

export const tests = import.meta.main ? null : {
  "should enable suggestions and list": ["foo", "\n"],
};

if (import.meta.main) {
  await Input.prompt({
    message: "Whats your name?",
    default: "foo",
    suggestions: ["foo", "bar", "baz"],
    list: true,
  });
}
