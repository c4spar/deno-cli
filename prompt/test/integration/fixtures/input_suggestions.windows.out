 ? Whats your name? (foo) » 
   foo
   bar
   baz
\x1b[1A\x1b[3A\x1b[29G\x1b[G\x1b[0J ? Whats your name? (foo) » foo
\x1b[?25h\x1b[?25h